### PR TITLE
Fix error interpolating field on irregularly spaced grid

### LINKED ISSUE #2911: Error interpolating on an irregularly spaced grid
Interpolating a field on an irregulalry spaced grid returns the wrong results. 

MWE:
```julia
julia> using Oceananigans
julia> using Oceananigans: interpolate

julia> z_faces(k) = 100 * ((k/Nz) ^ 2 - 2)
julia> grid = RectilinearGrid(topology = (Flat, Flat, Bounded), size = (Nz, ), z = z_faces)

julia> model = NonhydrostaticModel(; grid, tracers = :P)

julia> P(x, y, z) = z
julia> set!(model; P)

julia> interpolate(model.tracers.P, 0, 0, -100.0)
-118.95
```

### PR DESCRIPTION
**Pull request by jagoosw**
Fixes #2911

### CODE REVIEW

**Comment by jagoosw** [comment]
This is still not right

**Comment by simone-silvestri** [comment]
a `-1` on these lines should do it
https://github.com/CliMA/Oceananigans.jl/blob/20131e6a870185e3e49b9b0331870690cc3f7613/src/Fields/interpolate.jl#L57-L59

then we should make sure that also on `Face`s the interpolation is correct

**Comment by jagoosw** [comment]
@simone-silvestri that worked and looks correct on faces too

**Comment by navidcy** [comment]
How about a test?

**Comment by jagoosw** [comment]
Added definitions for fractional indices for Flat dimensions as suggested by @simone-silvestri [here](https://github.com/CliMA/Oceananigans.jl/issues/2911)

File: .buildkite/pipeline.yml
Comment (glwagner) [inline .buildkite/pipeline.yml:510]: Does this PR touch any code that affects this regression test? @simone-silvestri can you comment?

**Comment by simone-silvestri** [comment]
No it should not

**Comment by jagoosw** [comment]
Can I merge this?

File: test/test_field.jl
Comment (glwagner) [inline test/test_field.jl:113]: If you can please avoid `allowscalar`

Comment (jagoosw) [reply test/test_field.jl:113]: Is there a different way I can do this for the test without writing a kernel to evaluate `f`?

**Comment by navidcy** [comment]
Let's hold on merging until #2922 is resolved.

**Comment by navidcy** [comment]
can we resolve the conflicts?

**Comment by glwagner** [comment]
What's the status of this?

**Comment by jagoosw** [comment]
It was working just waiting for the shallow water test to be fixed. Some merge conflicts now which I can fix tomorrow.

**Comment by tomchor** [comment]
> @glwagner I'm not sure whats causing these tests to fail now?

afaik sometimes we get errors of the kind

`ERROR: LoadError: SystemError: opening file "/data5/glwagner/.julia-10703/compiled/v1.8/Oceananigans/hU93i_Y0P9A.ji": No such file or directory`

on the gpu tests. Which is what's hapenning now. It's now clear to me at least what the cause is, but restarting the tests generally makes them pass. I just restarted them so it be fine.

I should note that we got an error of the kind

 `Expression: all(test_fields.v .≈ truth_fields.v)`

in the GPU shallow water tests, which I thought were corrected. Is this something we need to worry about?

**Comment by jagoosw** [comment]
Can I merge now tests have passed?

**Comment by simone-silvestri** [comment]
sounds good to me!

**Comment by navidcy** [comment]
update/merge main first?
I see some file changes that they shouldn't be there..

**Comment by navidcy** [comment]
omg, I merged main and I noticed that the file changes in buildkite bringing back the SWE regression remained! it was commented out in main! I am sure I uncommented that at some point...

anyway, given that tests pass we should merge this

File: test/test_field.jl
Comment (navidcy) [inline test/test_field.jl:112]:
```suggestion
    # of f over the domain of the grid.

```

File: test/test_field.jl
Comment (navidcy) [inline test/test_field.jl:352]: this grid does not use the `FT`; was this change on purpose?

```suggestion
            reg_grid = RectilinearGrid(arch, FT, size=(4, 5, 7), x=(0, 1), y=(-π, π), z=(-5.3, 2.7), halo=(1, 1, 1))

```

**Comment by navidcy** [comment]
@jagoosw I committed a few changes. Feel free to undo or comment!

**Comment by jagoosw** [comment]
> @jagoosw I committed a few changes. Feel free to undo or comment!

Looks good, thank you!

**Comment by simone-silvestri** [comment]
> I think the test is failing becaue `f` is returning a `Float64`, so when the grid is `Float32` `eps(maximum(f))` is wrong. Is there a way to set the type returned by `f`?

you can also pass the grid
```julia
f(x, y, z, grid) = eltype(grid)(...)
ε_max = CUDA.@allowscalar f.(nodes(grid, (Face(), Face(), Face()), reshape=true)..., Ref(grid)) |> maximum |> eps
```

**Comment by jagoosw** [comment]
I don't understand why this is failing (it also fails locally).

It is only:
https://github.com/CliMA/Oceananigans.jl/blob/4d30ed8bf093b6b7894677ef33c2c0b1cf3e1487/test/test_field.jl#L126
that fails for Float32 on the grid points, and the error is ~4.6f-6 vs eps which is ~3.8f-6. It is also weird that this fails on regular grids where I haven't changed anything.

Also, the maximum error for each interpolation is:
```
u -> 1.9073486e-6
v ->  4.7683716e-6
w ->  3.8146973e-6
c -> 3.8146973e-6
```
with $\epsilon_{max} = 3.8146973f-6$ and I don't know why they wouldn't all be the same?

**Comment by glwagner** [comment]
> I don't understand why this is failing (it also fails locally).
> 
> It is only:
> 
> https://github.com/CliMA/Oceananigans.jl/blob/4d30ed8bf093b6b7894677ef33c2c0b1cf3e1487/test/test_field.jl#L126
> 
> 
> that fails for Float32 on the grid points, and the error is ~4.6f-6 vs eps which is ~3.8f-6. It is also weird that this fails on regular grids where I haven't changed anything.
> Also, the maximum error for each interpolation is:
> 
> ```
> u -> 1.9073486e-6
> v ->  4.7683716e-6
> w ->  3.8146973e-6
> c -> 3.8146973e-6
> ```
> 
> with ϵmax= 3.8146973f-6 and I don't know why they wouldn't all be the same?

I'm not sure what you mean by "they", but with different nodes, ie different `nodes(v, reshape=true)`, then shouldn't they be different (to within machine precision)?

The function involves a few arithmetic operations:

https://github.com/CliMA/Oceananigans.jl/blob/4d30ed8bf093b6b7894677ef33c2c0b1cf3e1487/test/test_field.jl#L109

and round off error accumulates. Therefore you may want to increase the tolerance to `10*e_max` or something.

**Comment by jagoosw** [comment]
> and round off error accumulates. Therefore you may want to increase the tolerance to `10*e_max` or something.

Ah I see, thank you, I wasn't sure how this worked. It passes locally now.

**Comment by jagoosw** [comment]
I can't re-run the tests but if someone with access can I think this PR should be done?

**Comment by jagoosw** [comment]
Not sure why the poisson test has failed now...

File: test/test_field.jl
Comment (navidcy) [reply test/test_field.jl:115]: Sure! Let’s fix it!

Comment (navidcy) [reply test/test_field.jl:115]: I don’t know how tho..

Comment (jagoosw) [reply test/test_field.jl:115]: I think the `nodes` line is fine here and it's just `f` line but not sure why

Comment (glwagner) [reply test/test_field.jl:115]: My suggested change above adds a comment to indicate the issue for future developers. Please change the comment if it is not clear (it seems like it isn't)!

Comment (jagoosw) [reply test/test_field.jl:115]: Oh I see, I didn't notice the comment before!

Comment (glwagner) [reply test/test_field.jl:116]: @navidcy this can be fixed by returning `SubArray` for "interior" nodes (not `OffsetArray`)

**Comment by jagoosw** [comment]
Can I merge now the tests pass?

**Comment by navidcy** [comment]
Yeah!

**Comment by navidcy** [comment]
Hm. Perhaps someone else might have other opinion?

**Comment by navidcy** [comment]
I think it’s done but I haven’t payed attention to what was happening here. I was excited to see the tests pass tho :)

File: src/Fields/interpolate.jl
Comment (glwagner) [reply src/Fields/interpolate.jl:73]: Why not replace `RectilinearGrid` with `AbstractGrid`?

Comment (glwagner) [reply src/Fields/interpolate.jl:73]: Or perhaps define a function like `isxregular` in `Grids`. That's probably better. This code should also be close to `RectilinearGrid`, its intimately tied to the specific implementatino of `RectilinearGrid` (ordering of type parameters etc)

File: src/Fields/interpolate.jl
Comment (jagoosw) [reply src/Fields/interpolate.jl:63]: Didn't mean to, I think this is missing bits of the new code when I was merging, will change!

File: src/Fields/interpolate.jl
Comment (jagoosw) [reply src/Fields/interpolate.jl:73]: I'm not sure this part actually needs to worry about flat topologies here since flat directions are delt with further down (and the topology is a parameter of `AbstractGrid`), so here what we want is just the same as `XRegLatLonGrid` which are defined:
https://github.com/CliMA/Oceananigans.jl/blob/9637514fa10334e25e3cd730b5eb28ec10e7e472/src/Grids/latitude_longitude_grid.jl#L77-L80

So I will replicate for `RectilinearGrid`

**Comment by glwagner** [comment]
> > I think the test is failing becaue `f` is returning a `Float64`, so when the grid is `Float32` `eps(maximum(f))` is wrong. Is there a way to set the type returned by `f`?
> 
> you can also pass the grid
> 
> ```julia
> f(x, y, z, grid) = eltype(grid)(...)
> ε_max = CUDA.@allowscalar f.(nodes(grid, (Face(), Face(), Face()), reshape=true)..., Ref(grid)) |> maximum |> eps
> ```
> 
> `eltype` is the function I couldn't remember!

I only saw this now -- just to help keep our tests well-written and easy to maintain you should use `convert` rather than the convenience constructors --- eg

```julia
f(x, y, z) = convert(typeof(x), ...)
```

File: src/Fields/interpolate.jl
Comment (glwagner) [reply src/Fields/interpolate.jl:73]: Can we avoid using these aliases at all and instead use a function `isxregular`?

The "alias" method + direct checking is annoying -- we have to update this code for new grids. The function-based approach is better.

Comment (glwagner) [reply src/Fields/interpolate.jl:73]: Can we please find a better design than using the aliases?

Comment (jagoosw) [reply src/Fields/interpolate.jl:73]: Is there a performance benefit to dispatching vs ifelse? If not it would probably look nicer to have `isregular` functions in an ifelse?

**Comment by jagoosw** [comment]
Think this is ready now?

**Comment by tomchor** [comment]
@jagoosw do you have merge permission? If so, I think you can go ahead and merge (unless anyone has further comments)

File: src/Grids/latitude_longitude_grid.jl
Comment (tomchor) [inline src/Grids/latitude_longitude_grid.jl:728]: Just a clarification about the nomenclature here. `LatLonGrid`s can be regular in longitude (i.e. when we're talking about degrees), but in that case Δx (i.e. in meters) can vary with latitude (even though they don't vary with longitude), no?

Do we still call that _regular_ x-spacing? Or do we somehow need to differentiate things a bit further?

File: src/Grids/Grids.jl
Comment (navidcy) [reply src/Grids/Grids.jl:142]: Good point about `iszregular`.

Comment (glwagner) [reply src/Grids/Grids.jl:142]: Probably `iszregular(grid) = true` is a safer fallback

File: src/Grids/latitude_longitude_grid.jl
Comment (glwagner) [reply src/Grids/latitude_longitude_grid.jl:728]: Where are these functions used? That should determine the correct return value. Then we can worry about the name.

File: src/Grids/Grids.jl
Comment (navidcy) [reply src/Grids/Grids.jl:142]: Safer would be not to have a fallback and define these for each grid. We only have 3 grids.

Comment (navidcy) [reply src/Grids/Grids.jl:142]: I had bad experiences with fallbacks...

Comment (jagoosw) [reply src/Grids/Grids.jl:142]: I went for falling back to `false` because the method for irregular grids should work even if the grid is regular (I think)?

File: src/Grids/latitude_longitude_grid.jl
Comment (jagoosw) [reply src/Grids/latitude_longitude_grid.jl:728]: In interpolation it is being assumed that we're talking about meters rather than longditude since we call `xspacings` and `xnodes` etc. which return in meters.

We probably need a way to tell if a user is interpolating in meters of degrees (although I don't think this should be a problem for how interpolation is used internally).

File: src/Grids/Grids.jl
Comment (glwagner) [reply src/Grids/Grids.jl:142]: i support no fallback. Grids are rare

File: src/Grids/latitude_longitude_grid.jl
Comment (simone-silvestri) [reply src/Grids/latitude_longitude_grid.jl:728]: Just recently the definition of `xnode` has changed for the `LatitudeLongitudeGrid` from degrees to meters. The particle advection assumes that the `x`, `y` coordinates of the particles are in degrees for a `LatitudeLongitudeGrid`

Comment (jagoosw) [reply src/Grids/latitude_longitude_grid.jl:728]: I think it would be most straight forward if the particles were always in meters? Otherwise we need to somehow detect in the interpolaiton if the user wants to interpolate in degrees or meters and I don't think interpolation is used anywhere else where it might be in degrees?

Comment (jagoosw) [reply src/Grids/latitude_longitude_grid.jl:728]: Changing particles in this PR as well as #2928 will be painful so it would probably be easiest to wait and change after thats merged

File: src/Grids/Grids.jl
Comment (jagoosw) [reply src/Grids/Grids.jl:142]: Should I get rid of the fallback for `isrectilinear` as well?

File: src/Grids/latitude_longitude_grid.jl
Comment (jagoosw) [reply src/Grids/latitude_longitude_grid.jl:728]: Its should be quite easy to change to Lagrangian particles being in meters since we would just need to remove the `corrdinate_transform_u` etc., it might be more challenging to change the `return_face_metrics` function

Comment (jagoosw) [reply src/Grids/latitude_longitude_grid.jl:728]: On the other hand, changing to work in meters will make most `LatitudeLongditudeGrid` particle simulaitons much slower because most will have regular lat/lon but not meters so it probably would be better to stick with lat/lon?

**Comment by jagoosw** [comment]
I think I've fixed this for lat/lon grids given the new `x/ynode` definitions for lat/lon grids without touching the lagrangian particles (and assuming they're still positioned in degrees not meters) (@tomchor)

**Comment by glwagner** [comment]
> I think I've fixed this for lat/lon grids given the new `x/ynode` definitions for lat/lon grids without touching the lagrangian particles (and assuming they're still positioned in degrees not meters) (@tomchor).
> 
> I'm not so keen on my function names:
> 
> https://github.com/CliMA/Oceananigans.jl/blob/92709dd96852af2806ec91148e6d463c67f25787/src/Fields/interpolate.jl#L61-L77
> 
> 
> but can't think of better alternatives

I tthink you just need

```julia
node(1, j, k, grid, loc...)[1]
```

So you don't need special function names.

**Comment by glwagner** [comment]
It looks like none of this stuff will work on `OrthogonalSphericalShellGrid` because we assume that the spacings are functions of one index only, rather than `i, j`. Is that true?

**Comment by jagoosw** [comment]
Yeah, I guess this would all be a problem on `OrthogonalSphericalShellGrid` and would all need changing. And yeah I think the API is inconsistent for `loc`, sometimes it is expecting a tuple of locs and other times it is expecting each component separatly

**Review by tomchor — COMMENTED** [review]
Sorry @jagoosw I kinda lost track of the discussion regarding nomenclature, units, etc. Was all that resolved in a way that everybody's happy with?

Otherwise I think the code looks good!

**Comment by jagoosw** [comment]
I think it was. We've stuck with degrees for particles on lat/lon grids. I think the only outstanding question is perhaps the issues with `OrthogonalSphericalShellGrid` but particles don't work on them and that's the only place this interpolation gets used.

**Comment by tomchor** [comment]
> I think it was. We've stuck with degrees for particles on lat/lon grids. I think the only outstanding question is perhaps the issues with `OrthogonalSphericalShellGrid` but particles don't work on them and that's the only place this interpolation gets used.

Took another look at the code. In my opinion this is ready to merge.

File: src/Fields/interpolate.jl
Comment (glwagner) [inline src/Fields/interpolate.jl:76]: Can we write this in more lines? I can't read it.

File: src/Fields/interpolate.jl
Comment (glwagner) [inline src/Fields/interpolate.jl:95]: Same as above, needs to be written out

File: test/test_field.jl
Comment (glwagner) [inline test/test_field.jl:364]: i think we've agreed to call this a "stretched grid" rather than an irregular grid.

File: test/test_field.jl
Comment (glwagner) [inline test/test_field.jl:362]:
```suggestion
                                         z = [-5.3, -4.2, -3.0, -1.9, -0.7, 0.4, 1.6, 2.7], halo=(1, 1, 1))

```

File: src/Fields/interpolate.jl
Comment (glwagner) [reply src/Fields/interpolate.jl:76]: Something like

```julia
Tx = topology(grid, 1)
Nx = size(grid, 1)
L = length(loc, Tx(), Nx)
xn = nodes(grid, locs)[1]
return fractional_index(L, x, xn) - 1
```

File: src/Fields/interpolate.jl
Comment (glwagner) [inline src/Fields/interpolate.jl:61]: I dont think `FT` is used in this function. Also it might be better to use `x = nodes(grid, locs)[1]` and change the name of the input to something like `xi`.

Comment (glwagner) [reply src/Fields/interpolate.jl:61]: true, I missed that!

**Review by glwagner — APPROVED** [review]
great work

**Comment by jagoosw** [comment]
Please could someone re-run the GPU tests?

**Comment by simone-silvestri** [comment]
the /tmp directory in sverdrup is full. When we have cleaned it up we can rerun

**Comment by jagoosw** [comment]
Can I merge?

**Comment by simone-silvestri** [comment]
go for it!

**Pull request merged**
